src/lib/Persistence/Legacy/Token/Gateway/Token/Doctrine/DoctrineGateway.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Core\Persistence\Legacy\Token\Gateway\Token\Doctrine;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Ibexa\Core\Base\Exceptions\NotFoundException as NotFound;
+use Ibexa\Core\Persistence\Legacy\Token\AbstractGateway;
+use Ibexa\Core\Persistence\Legacy\Token\Gateway\Token\Gateway;
+use Ibexa\Core\Persistence\Legacy\Token\Gateway\TokenType\Doctrine\DoctrineGateway as TokenTypeGateway;
+
+/**
+ * @internal
+ */
+final class DoctrineGateway extends AbstractGateway implements Gateway
+{
+    public const TABLE_NAME = 'ibexa_token';
+    public const DEFAULT_TABLE_ALIAS = 'token';
+
+    public const COLUMN_ID = 'id';
+    public const COLUMN_TYPE_ID = 'type_id';
+    public const COLUMN_TOKEN = 'token';
+    public const COLUMN_IDENTIFIER = 'identifier';
+    public const COLUMN_CREATED = 'created';
+    public const COLUMN_EXPIRES = 'expires';
+    public const TOKEN_SEQ = 'ibexa_token_id_seq';
+
+    private Connection $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public static function getColumns(): array
+    {
+        return [
+            self::COLUMN_ID,
+            self::COLUMN_TYPE_ID,
+            self::COLUMN_TOKEN,
+            self::COLUMN_IDENTIFIER,
+            self::COLUMN_CREATED,
+            self::COLUMN_EXPIRES,
+        ];
+    }
+
+    public function insert(
+        int $typeId,
+        string $token,
+        ?string $identifier,
+        int $ttl
+    ): int {
+        $now = $this->getCurrentUnixTimestamp();
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->insert(self::TABLE_NAME)
+            ->values([
+                self::COLUMN_TYPE_ID => ':type_id',
+                self::COLUMN_TOKEN => ':token',
+                self::COLUMN_IDENTIFIER => ':identifier',
+                self::COLUMN_CREATED => ':created',
+                self::COLUMN_EXPIRES => ':expires',
+            ])
+            ->setParameter(':type_id', $typeId, ParameterType::INTEGER)
+            ->setParameter(':token', $token, ParameterType::STRING)
+            ->setParameter(':identifier', $identifier, ParameterType::STRING)
+            ->setParameter(':created', $now, ParameterType::INTEGER)
+            ->setParameter(':expires', $now + $ttl, ParameterType::INTEGER);
+
+        $query->execute();
```

Comment:
You can use `Connection::insert` method directly. Query builder is not necessary here.

```suggestion
        $this->connection->insert(
            self::TABLE_NAME,
            [
                self::COLUMN_TYPE_ID => $typeId,
                self::COLUMN_TOKEN => $token,
                self::COLUMN_IDENTIFIER => $identifier,
                self::COLUMN_CREATED => $now,
                self::COLUMN_EXPIRES => $now + $ttl,
            ], [
                self::COLUMN_TYPE_ID => ParameterType::INTEGER,
                self::COLUMN_CREATED => ParameterType::INTEGER,
                self::COLUMN_EXPIRES => ParameterType::INTEGER,
            ]
        );
```

Note that this also makes use of parameterized query, so there is no SQL injection attack vector.

Additionally I skipped the string parameter type declarations, because those are the default.